xc/device.ml

```diff
@@ -2227,6 +2231,79 @@ module Backend = struct
       let cmdline_of_info ~xs ~dm info restore domid =
         let common = Dm_Common.cmdline_of_info ~xs ~dm info restore domid ~domid_for_vnc:true in
 
+        let usb =
+        match info.Dm_Common.usb with
+        | Dm_Common.Disabled -> []
+        | Dm_Common.Enabled devices ->
+          let devs = devices |>
+            List.map (fun (x,y) -> ["-device"; sprintf "usb-%s,port=%d" x y]) |> List.concat in
+          "-usb" :: devs
+        in
+
+        let serial_device =
+          try
+            let xs_path = xs.Xs.read "/local/logconsole/@" in
+            let file =
+            if Stdext.Xstringext.String.has_substr xs_path "%d" then
+              Stdext.Xstringext.String.replace "%d" (string_of_int domid) xs_path
+            else
+              xs_path
+            in Some ("file:" ^ file)
+          with _ -> info.Dm_Common.serial
+        in
+
+        let list1 = ["piix3-ide-xen"; "piix3-usb-uhci"; "rtl8139"] in
+        let list2 = ["subvendor_id=0x5853"; "subsystem_id=0x0001"] in
+        let global = List.map (fun x -> List.map (fun y -> x^"."^y) list2) list1 |> List.concat in
+
+        let qmp = ["libxl"; "event"] |>
+          List.map (fun x -> ["-qmp"; sprintf "unix:/var/run/xen/qmp-%s-%d,server,nowait" x domid]) |>
+          List.concat in
+
+        let has_platform_device =
+          try
+            int_of_string (xs.Xs.read (sprintf "/local/domain/%d/vm-data/disable_pf" domid)) <> 1
```

Comment:
Could we compare just against a string `"1"` without conversion? Is there danger that the string contain additional spaces? The original script uses conversion - so let's keep it.

Reply:
If the string contains spaces the try-catch block will return true either we compare `" 1"` and `"1"` as strings or we convert them since int_of_string throws an exception - or should I trim what I read from xenstore? Is it likely to have spaces? I would think a value padded left with zeros might be more likely?